File: client/src/stores/rooms.ts
import { createSignal } from "solid-js";
import { invoke } from "@tauri-apps/api/core";

export interface Room {
  id: string;
  name: string;
  description: string | null;
  member_count: number;
  owner_id: string;
  invite_code?: string;
}

export interface RoomSettings {
  id: string;
  name: string;
  description: string | null;
  owner_id: string;
  member_count: number;
  invite_code?: string;
  members: RoomMember[];
}

export interface RoomMember {
  user_id: string;
  display_name: string;
  role: "owner" | "member";
}

export interface RoomResult {
  ok: boolean;
  error?: string;
  room?: Room;
  invite_code?: string;
}

const [rooms, setRooms] = createSignal<Room[]>([]);

export { rooms };

export async function fetchRooms(): Promise<void> {
  try {
    const list = await invoke<Room[]>("get_rooms");
    setRooms(list);
  } catch {
    // silently keep stale list
  }
}

export async function createRoom(
  name: string,
  description: string | undefined,
): Promise<RoomResult> {
  try {
    const room = await invoke<Room>("create_room", { name, description: description ?? "" });
    setRooms((prev) => [...prev, room]);
    return { ok: true, room };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function joinByCode(code: string): Promise<RoomResult> {
  try {
    const room = await invoke<Room>("join_by_code", { code });
    setRooms((prev) => {
      if (prev.some((r) => r.id === room.id)) return prev;
      return [...prev, room];
    });
    return { ok: true, room };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function leaveRoom(roomId: string): Promise<RoomResult> {
  try {
    await invoke("leave_room", { roomId });
    setRooms((prev) => prev.filter((r) => r.id !== roomId));
    return { ok: true };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function getRoomSettings(roomId: string): Promise<RoomResult & { room?: RoomSettings }> {
  try {
    const settings = await invoke<RoomSettings>("get_room_settings", { roomId });
    return { ok: true, room: settings };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function updateRoom(
  roomId: string,
  changes: { name: string; description?: string },
): Promise<RoomResult> {
  try {
    const { name, description } = changes;
    await invoke("update_room", { roomId, name, description: description ?? "" });
    setRooms((prev) =>
      prev.map((r) => (r.id === roomId ? { ...r, name, description: description ?? null } : r))
    );
    return { ok: true };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function deleteRoom(roomId: string): Promise<RoomResult> {
  try {
    await invoke("delete_room", { roomId });
    setRooms((prev) => prev.filter((r) => r.id !== roomId));
    return { ok: true };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

export async function regenerateInvite(roomId: string): Promise<RoomResult> {
  try {
    const code = await invoke<string>("regenerate_invite", { roomId });
    return { ok: true, invite_code: code };
  } catch (e: unknown) {
    return { ok: false, error: String(e) };
  }
}

File: client/src-tauri/src/commands/rooms.rs
use serde::{Deserialize, Serialize};
use tauri::State;

use crate::http_client::HttpClient;
use crate::state::AppState;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Room {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub member_count: i64,
    pub owner_id: String,
    pub invite_code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomSettings {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub owner_id: String,
    pub member_count: i64,
    pub invite_code: Option<String>,
    pub members: Vec<RoomMember>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomMember {
    pub user_id: String,
    pub display_name: String,
    pub role: String,
}

#[tauri::command]
pub async fn get_rooms(state: State<'_, AppState>) -> Result<Vec<Room>, String> {
    let http = HttpClient::new();
    let req = http.sig_get(&state, "/rooms").await?;
    http.send_json(req).await
}

#[tauri::command]
pub async fn create_room(
    name: String,
    description: String,
    state: State<'_, AppState>,
) -> Result<Room, String> {
    let http = HttpClient::new();
    let req = http
        .sig_post(&state, "/rooms")
        .await?
        .json(&serde_json::json!({
            "name": name,
            "description": description,
        }));
    http.send_json(req).await
}

#[tauri::command]
pub async fn join_by_code(code: String, state: State<'_, AppState>) -> Result<Room, String> {
    let http = HttpClient::new();
    let req = http
        .sig_post(&state, "/rooms/join")
        .await?
        .json(&serde_json::json!({ "invite_code": code }));
    http.send_json(req).await
}

#[tauri::command]
pub async fn leave_room(room_id: String, state: State<'_, AppState>) -> Result<(), String> {
    let http = HttpClient::new();
    let path = format!("/rooms/{room_id}/leave");
    let req = http.sig_post(&state, &path).await?;
    http.send_empty(req).await
}

#[tauri::command]
pub async fn get_room_settings(
    room_id: String,
    state: State<'_, AppState>,
) -> Result<RoomSettings, String> {
    let http = HttpClient::new();
    let path = format!("/rooms/{room_id}");
    let req = http.sig_get(&state, &path).await?;
    http.send_json(req).await
}

#[tauri::command]
pub async fn update_room(
    room_id: String,
    name: String,
    description: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let http = HttpClient::new();
    let path = format!("/rooms/{room_id}");
    let req = http.sig_put(&state, &path).await?.json(&serde_json::json!({
        "name": name,
        "description": description,
    }));
    http.send_empty(req).await
}

#[tauri::command]
pub async fn delete_room(room_id: String, state: State<'_, AppState>) -> Result<(), String> {
    let http = HttpClient::new();
    let path = format!("/rooms/{room_id}");
    let req = http.sig_delete(&state, &path).await?;
    http.send_empty(req).await
}

#[tauri::command]
pub async fn regenerate_invite(
    room_id: String,
    state: State<'_, AppState>,
) -> Result<String, String> {
    let http = HttpClient::new();
    let path = format!("/rooms/{room_id}/invite");
    let req = http.sig_post(&state, &path).await?;

    #[derive(Deserialize)]
    struct InviteResp {
        invite_code: String,
    }
    let resp: InviteResp = http.send_json(req).await?;
    Ok(resp.invite_code)
}
